\documentclass{standalone}

\begin{document}

\section[CHIMeRA]{The CHIMeRA project}\label{chimera:chimera}

The increasing availability of large-scale biomedical literature under the form of public on-line databases, has opened the door to a whole new understanding of multi-level associations between genomics, protein interactions and metabolic pathways for human diseases.
Many structures and resources aiming at such type of analyses have been built, with the purpose of disentangling the complex relationships between various aspects of the human system, relating to diseases~\cite{SymtomsNet, HumanPhenotype, chimerDB2017}.
Such structures, while allowing to study disease-to-some-other-omic associations, may not be sufficient when trying to bridge the gap of interpreting results and concept proofing clinical studies, when many types of data are involved.
Looking for causation of diseases across different omics has also became a major challenge, with the aim of expanding etiology and obtaining insights on pathogenesis~\cite{Barabasi2007}.
This task may prove to be particularly hard when dealing with medical ontology strings, coming from different sources.
Information of this type are usually provided by brief sentences and periphrases, while synonyms may occur to describe the same concepts, thus causing different data source to provide different relationships for similar instances.
Text mining and string processing, thus is becoming a required step when trying to exploit medical ontology as a bridge to diffuse information.

All these data come from different kinds of studies, performed by independent research groups, who want to prove their theory about a particular aspect of biological agent interactions.
Modern biological analyses perform very capillary studies on biomedical agents, deeply studying the relationships between them but loosing information about what there is around them.
This approach is extremely efficient for the detection of the minimal causal agents of a problem, but it tends to loose its global and complex\footnote{
  From a physical point-of-view.
} behavior.
This is starting point of complex systems, i.e systems composed by multiple components with a mutual interactions between them.
The study of an individual aspect, in fact, could give us only a partial overview of the system, but we have to take into account the interactions between the multiple components for a global description.

Network structures are acquiring even more importance on this kind of studies.
Complex System and System Biology researches have proposed multiple models about the dynamical and evolutionary interactions of the human system agents, aiming to study the hidden relationships between them using graph models.
A network structure, in fact, is able to highlight and quantify non-trivial correlations between system components.
The mathematical definition of network structures is given by the Graph Theory.
We define a network/graph as a pair $G = (V, E)$, where $V$ is a set of elements called nodes (or vertices) and $E$ is the set of their pairwise associations (links or edges).
The total number of graph nodes (or cardinality of the graph) is denoted by $N$ and it defines the order of the graph.
The graph dimension is given by the number of its edges ($m$).
We define a graph as \emph{complete graph} if it has all its possible edges ($m = N \times N$).
A network made by nodes of the same type could be described via its adjacency matrix, i.e a matrix $(N\times N)$ in which each row/column identifies a node and each link $e_{ij}$ quantifies the importance of the interaction between node $v_i$ and node $v_j$.
We define the importance of a node into the graph using the number of its connections: this is a classical \emph{centrality measure} and it is called node \emph{degree} centrality.
Starting from these definitions, we can enrich our model combining multiple network structure: given two graphs $G(V, E)$ and $G'(V', E')$, we define their combination as a new graph, where its nodes are given by the intersection of $V\cap V'$ and its edges are given by $E\cap E'$.
If $V\cap V'=\emptyset$ the two graphs are \emph{disjointed}; in contrary, if $V'\subseteq V$ and $E'\subseteq E$ then $G'$ is a subgraph of $G$.
Combining multiple graphs together, possibly including nodes of different types, we obtain a network-of-networks structure able to map a wide range of interactions from multiple sets of elements.
We will describe its properties later.

In real data applications, we can often reasonably assume that a wide amount of matrix entries are null, i.e the interaction between the involved agents is quite sparse, and we can use important properties related to sparse matrices to manage our network.
However, when the amount of data increases, also the management of a such sparse matrix could be difficult.
More efficient solutions are provided by modern Database formats and languages (e.g \textsf{MySQL}, \textsf{SQLite}, \textsf{InfluxDB}, $\cdots$), which store all the information into a binary format and they allow to submit queries to extract the desired portion of data.
A global visualization of these huge amount of data is, in fact, without practical-sense and none valuable information can be extracted from the global representation of the system.
The most important feature of network model is, in fact, the definition of a hierarchy of interactions: the relationship between two nodes is given by the amount of connections which link them or, in other words, by their paths.
Starting from a node, its nearest neighbors are given by the set of nodes connected to it: re-iterating this concept we can explore all the network structure\footnote{
  We assume that our network structure has not isolated nodes and it has only undirected connections.
}.
In this way we can study the interactions of each node at different precision orders and causalities.

In light of these considerations, we started to develop the \textsf{CHIMeRA} project (\emph{Complex Human Interactions in MEdical Records and Atlases}), where we aim to merge state-of-art studies and databases about biomedical researches into a unified network structure.
A key role on our network structure is played by diseases: the major part of biomedical researches are focused on causes and consequences of a given disease, involving the corresponding databases to store the interactions between them and other biological factors.
Diseases are also the most bigger manifestations of biological malfunctions and a large part of biomedical researches are financed on their study, looking for their fine grain causes.
Thus, a disease could be a valid \quotes{bridge} between multiple data sources: merging disease nodes derived from different datasets we can provide a unique structure which host all the information.

The crucial point of this project has been, in fact, the merging of different kind of information provided by multiple distinct data structured.
As told above, the major part of researches have focused on a partial aspect of the problem and they provide an independent result from the others, reducing the possibility of interactions between the outputs.
We tend to spend a lot of computational power to visualize the results using web pages and on-line services, but they drastically affects the real usage of these information.
The \textsf{CHIMeRA} project has began from these independent sources, aiming to maximize their overlap and, thus, the communications between them.

We have to pay a final attention about the format of these data: in physics we are friendly with numerical data, but in these context we have to work with words and text strings.
The above told databases include only the research outputs and, thus, the interpretations of analyzed numerical data.
For example, if a numerical significant correlation was found between a disease and a gene we would find an association between them into a database (modeled as a link in our network).
The only information available into this database is a link between two words, the disease name and the gene name, without any numeric value.
While numbers have a unique representation (the number 42 is always 42) we can use multiple periphrases, i.e set of strings, to identify the same concept.
The biomedical community, in fact, has not yet provided a (public\footnote{
  For sake of completeness we have to mention the \href{https://www.meddra.org/}{MedDRA} database which is a pay-to-use repository of these information.
}) unified standard for disease identification or, at least, it has not yet provided a rigid standard as for other kind of data as genes or SNPs.
So, if diseases could be an efficient way to link together multiple data sources, they throwback an extreme variability in their nomenclature.
The \textsf{CHIMeRA} project has tried to overcome this issue, using a Natural Language Processing (NLP) approach.

In the next sections we are going to discuss about the multiple steps which lead us to the formulation of our unified \textsf{CHIMeRA} database.
We will start from the preliminary studies performed in relation to the INFN FiloBlu project, which allow the creation of the \textsf{SymptomsNet} structure, i.e a \quotes{smaller} network based on Italian words which links diseases to related symptoms.
Then we will briefly introduce the most common NLP techniques, also used into the \textsf{CHIMeRA} pipeline and, finally, we will show the main developed features of our \textsf{CHIMeRA} network.

%What is CHIMeRA project and which is its potentiality.
%Description of the database created and of the query implemented to obtain the results


\end{document}